src/utils/hierarchical_model.py

```diff
@@ -0,0 +1,213 @@
+        pd.Series(str): A list of colors, where the i-th element corresponds to the color of the i-th row in the DataFrame.
+    """
+    # Get the unique cluster labels from the DataFrame
+    cluster_labels = np.unique(labels)
+
+    # Assign a color to each cluster label
+    cluster_colors = {label: colors_code[i] for i, label in enumerate(cluster_labels)}
+    
+    # Create a new column in the DataFrame to store the cluster colors
+    return pd.Series(labels).apply(lambda x: cluster_colors[x])
+
+
+def get_transform_PCA(data: pd.DataFrame, dimensions: int) -> ndarray:
+    """Performs principal component analysis (PCA) on a given pandas DataFrame data
+        and returns the transformed data.
+
+    Args:
+        data (pd.DataFrame): The input data to be transformed using PCA.
+        dimensions (int): The number of dimensions to keep after the transformation.
+    Returns:
+        ndarray: A numpy ndarray of shape (n_samples, dimensions) representing the transformed data.
+    """
+    # Perform PCA on the data
+    pca = PCA(n_components=dimensions)
+    return pca.fit_transform(data)
+
+
+def plot_clustering_before_reduction(data: pd.DataFrame, labels:ndarray, dimensions: int) -> None:
+    """Generates a scatter plot of the clustering of data before reducing its dimensionality using PCA.
+
+    Args:
+        data (pd.DataFrame): The input dataset to be plotted.
+        dimensions (int): The number of dimensions to reduce the data to before plotting.
+    Raises:
+        ValueError: To check the requested dimensions are either 2 or 3.
+    """
+    # Check the requested dimensions
+    if dimensions not in [2, 3]:
+        raise ValueError("dimensions must be either 2 or 3")
+    
+    # Perform PCA on the data
+    transformed = get_transform_PCA(data, dimensions)
+    
+    # Assigns a color to each cluster in the given DataFrame based on their label.
+    colors = _map_color_cluster(labels)
+    
+    # Set up plot
+    fig = plt.figure()
+    
+    # Plot scatter by PCA transformed data
+    if dimensions == 2:
+        ax = fig.add_subplot(111)
+        ax.scatter(transformed[:, 0], transformed[:, 1], c=colors)
+    else:
+        ax = fig.add_subplot(111, projection='3d')
+        ax.scatter(transformed[:, 0], transformed[:, 1], transformed[:, 2], c=colors)
+        ax.set_zlabel('PC3')
+
+    # Set axis labels and title
+    ax.set_title(f'{dimensions}-Dimension Clustering before PCA')
+    ax.set_xlabel('PC1')
+    ax.set_ylabel('PC2')
+    
+    # Show the plot
+    plt.show()
+
+
+def plot_clustering_after_reduction(labels: ndarray, transformed: ndarray, dimensions: int) -> None:
+    """Generates a scatter plot of the clustering of data after reducing using the given transformed data by PCA.
+
+    Args:
+        data (pd.DataFrame): The input dataset to be plotted.
+        transformed (ndarray): The transformed data using PCA. 
+        dimensions (int): The number of dimensions to reduce the data to after plotting.
```

Comment:
I think this description is misleading did you mean number of components generated by the PCA?

Reply:
I update the documentations of 'hierarchical_model' file.